pkg/api/api.go

```diff
@@ -442,12 +454,91 @@ func equalASCIIFold(s, t string) bool {
 	return s == t
 }
 
+// newStamperPutter returns either a storingStamperPutter or a pushStamperPutter
+// according to whether the upload is a deferred upload or not. in the case of
+// direct push to the network (default) a pushStamperPutter is returned.
+// returns a function to wait on the errorgroup in case of a pushing stamper putter.
+func (s *server) newStamperPutter(r *http.Request) (storage.Storer, func() error, error) {
+	batch, err := requestPostageBatchId(r)
+	if err != nil {
+		return nil, noopWaitFn, fmt.Errorf("postage batch id: %w", err)
+	}
+
+	deferred, err := requestDeferred(r)
+	if err != nil {
+		return nil, noopWaitFn, fmt.Errorf("request deferred: %w", err)
+	}
+
+	if deferred {
+		p, err := newStoringStamperPutter(s.storer, s.post, s.signer, batch)
+		return p, noopWaitFn, err
+	}
+	p, err := newPushStamperPutter(s.storer, s.post, s.signer, batch, s.chunkPushC)
+	return p, p.eg.Wait, err
+}
+
+type pushStamperPutter struct {
+	storage.Storer
+	stamper postage.Stamper
+	eg      errgroup.Group
+	c       chan *pusher.Op
+	sem     chan struct{}
+}
+
+func newPushStamperPutter(s storage.Storer, post postage.Service, signer crypto.Signer, batch []byte, cc chan *pusher.Op) (*pushStamperPutter, error) {
+	i, err := post.GetStampIssuer(batch)
+	if err != nil {
+		return nil, fmt.Errorf("stamp issuer: %w", err)
+	}
+
+	stamper := postage.NewStamper(i, signer)
+	return &pushStamperPutter{Storer: s, stamper: stamper, c: cc, sem: make(chan struct{}, 20)}, nil
+}
+
+func (p *pushStamperPutter) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exists []bool, err error) {
+	exists = make([]bool, len(chs))
+
+	for i, c := range chs {
+		// skips chunk we already know about
+		has, err := p.Storer.Has(ctx, c.Address())
+		if err != nil {
+			return nil, err
+		}
+		if has || containsChunk(c.Address(), chs[:i]...) {
+			exists[i] = true
+			continue
+		}
+		stamp, err := p.stamper.Stamp(c.Address())
+		if err != nil {
+			return nil, err
+		}
+
+		func(ch swarm.Chunk) {
+			p.sem <- struct{}{}
+			p.eg.Go(func() error {
+				errc := make(chan error, 1)
+				// note: shutdown might be tricky, we need to pass the quit channel
```

Comment:
Would it be better to have a separate `chan` for each `pushStamperPutter`. Since the `smugler` can take in multiple channels, we can handle closing of each channel separately and handle the global channel closing better?

So for every upload we will try to enqueue a new `pusher.Op` channel to the smugler. So we can wait till the channel is enqueued till the request context is closed. Once it is enqueued, a separate routine will read from the channel uninterrupted. This way a naked write here would be okay.

Reply:
I think it is a nice idea, but do you think it makes a difference? you still need to invoke a goroutine for every chunk through the error group, and you still need to do the individual allocations for the channels here while just adding more complexity on the pusher side, no? I would create a `sync.Pool` of channels and just pool the individual operation channels, so that channel allocations become less wasteful. Also, the naked write should be replaced with a select case as you pointed out :+1: